ccmbioinfo/lfs
# CARDS-1874: CARDS should support connecting to external Mongo Replica Sets in addition to singular external Mongo instances

This PR adds support for MongoDB replica sets as a data storage engine for Jackrabbit Oak by passing the `mongo.uri` variable directly to `org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService` instead of as a mix of the variables; `mongo.host`, `mongo.port`, `mongo.db`, and `mongo.auth` which prevented the specification of a replica set connection string such as `mongo1:27017,mongo2:27017,mongo3:27017/?replicaSet=uhndatalake`.

Testing
---------

- Build this `CARDS-1874` branch with the `docker` profile enabled with `mvn clean install -Pdocker`.
- Start a test local MongoDB replica set from the https://github.com/IntegralProgrammer/local-mongo-replicaset repository,
- Verify that CARDS can use this MongoDB replica set by starting a CARDS instance with the command `./start_cards.sh --project cards4heracles --dev --mongo -V mongo.uri=root:password@mongo1:30001,mongo2:30002,mongo3:30003/?replicaSet=uhndatalake`.
- Ensure that it starts properly and that you are able to login and create test _Forms_ and _Subjects_.
- Stop CARDS with CTRL+C.
- `cd compose-cluster`
- `./cleanup.sh`
- `python3 generate_compose_yaml.py --dev_docker_image --composum --cards_project cards4heracles --subnet 172.99.0.0/16 --external_mongo --external_mongo_uri mongo1:30001,mongo2:30002,mongo3:30003/?replicaSet=uhndatalake`
- Enter `root:password` for the Mongo password when prompted
- Edit `docker-compose.yml` and add the following lines into the configuration for the `cardsinitial` service:

```yaml
extra_hosts:
- "mongo1:172.99.0.1"
- "mongo2:172.99.0.1"
- "mongo3:172.99.0.1"
```

- `docker-compose build`
- `docker-compose up -d`
- Visit http://localhost:8080 and login as `admin`:`admin`. The previously created _Forms_ and _Subjects_ should be there as the Docker Compose deployment uses the same MongoDB replica set that `start_cards.sh` used.

Clean up
------------

- Bring down the Docker Compose CARDS environment with `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
- Bring down the local MongoDB Replica Set as per https://github.com/IntegralProgrammer/local-mongo-replicaset#stopping
